### PR TITLE
zamjene sol

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/coci-16-zamjene.mdx [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+author: Kevin Sheng
+---
+
+[Official Analysis (C++)](https://github.com/mostafa-saad/MyCompetitiveProgramming/blob/master/Olympiad/COCI/official/2017/contest2_solutions)
+
+## Explanation
+
+If you've done [Wormhole Sort](http://usaco.org/index.php?page=viewproblem2&cpid=992),
+this problem should look a bit familiar.
+From that problem's editorial, we know that it's possible to sort a component
```
Comment (ryanchou-dev) [inline solutions/coci-16-zamjene.mdx:14]: this way of mentioning the editorial is iffy, but it's alright. if you can, summarize a little bit.

File: solutions/coci-16-zamjene.mdx [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+iff each positions' *sorted position* is in the same component as itself.
```
Comment (ryanchou-dev) [inline solutions/coci-16-zamjene.mdx:15]:
```suggestion
if each positions' *sorted position* is in the same component as itself.
```
iff 🥺

File: solutions/coci-16-zamjene.mdx [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+#include <iostream>
+#include <vector>
+#include <map>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+class DominikArray {
```
Comment (ryanchou-dev) [inline solutions/coci-16-zamjene.mdx:99]: codesnip pls

**Review by ryanchou-dev — APPROVED** [review]

File: solutions/coci-16-zamjene.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/coci-16-zamjene.mdx:15]: if & only if?

File: solutions/coci-16-zamjene.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/coci-16-zamjene.mdx:99]: this class is the meat of the entire program

File: solutions/coci-16-zamjene.mdx [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+we can obtain our hash by multiplying each power by the number of times it occurs in the set.
+
+For example, if our component was the multiset ${3, 3, 4, 4, 5}$, and each number's power was itself, our hash would be:
+$$
+2 \cdot P^3 + 2 \cdot P^4 + 1 \cdot P^5
+$$
+
+<Info>
+
+It's also possible to assign each distinct number in the array a random value from a large range like $[1,10^6]$.
```
Comment (bqi343) [inline solutions/coci-16-zamjene.mdx:45]: what I actually meant was that if you replace every element of `elem_pow` with a random 64-bit integer the solution should still work.

**Review by bqi343 — CHANGES_REQUESTED** [review]
there's no analysis of the collision probability

File: solutions/coci-16-zamjene.mdx (hunk 4 above)
Comment (bqi343) [reply solutions/coci-16-zamjene.mdx:45]: would be good to think about what the collision probability is if you do this

File: solutions/coci-16-zamjene.mdx (hunk 2 above)
Comment (ryanchou-dev) [reply solutions/coci-16-zamjene.mdx:15]: oops meant to format the change like this:
```suggestion
only if each positions' *sorted position* is in the same component as itself.
```

Comment (ryanchou-dev) [reply solutions/coci-16-zamjene.mdx:15]: up to u though, iff looks fine upon second inspection

File: solutions/coci-16-zamjene.mdx (hunk 3 above)
Comment (ryanchou-dev) [reply solutions/coci-16-zamjene.mdx:99]: oops im dumb ;w;

File: solutions/coci-16-zamjene.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/coci-16-zamjene.mdx:45]: > what I actually meant was that if you replace every element of `elem_pow` with a random 64-bit integer the solution should still work.

that's what i meant by this?

Comment (bqi343) [reply solutions/coci-16-zamjene.mdx:45]: "from a large range like $[1,10^6]$" is not the same as "a random 64-bit integer"

**Comment by SansPapyrus683** [comment]
chance of two hashes colliding w/ 2^64 is 1/2^64 right, just making sure

**Comment by bqi343** [comment]
this is false

**Comment by SansPapyrus683** [comment]
no clue what you mean by that

**Comment by bqi343** [comment]
Can you explain how you got 1/2^{64}? Obviously if you pick two random numbers from the range $[0,2^{64})$ this will be the probability they collide, but that's not what your solution is doing.

**Comment by SansPapyrus683** [comment]
i though that the 1/10^9 collision chance for a prime mod would generalize to 2^64

**Comment by SansPapyrus683** [comment]
what’s N?

**Comment by bqi343** [comment]
the length of the string; check the module

**Comment by SansPapyrus683** [comment]
so what's the chance for nonprime MODs?

**Comment by bqi343** [comment]
Well, for $2^{64}$ specifically there is no guarantee. Not sure about other nonprime MODs.

**Comment by SansPapyrus683** [comment]
so i should just put a disclaimer "this code is vulnerable to hacking bc it doesn't use a prime mod"?

**Comment by bqi343** [comment]
I think you should fix the solution *and* add a note to that effect

**Comment by SansPapyrus683** [comment]
what's wrong w/ the solution? it gets accepted just fine?

**Comment by bqi343** [comment]
But if the test data was stronger it wouldn't.

**Comment by SansPapyrus683** [comment]
can't that be said about basically any hashing problem lol
anyways fine, i'll use double hashing

**Comment by bqi343** [comment]
> can't that be said about basically any hashing problem lol

No. If your solution uses a sufficiently large prime mod and a random base then it will be unhackable.

On the other hand, there is a test case such that your solution fails for *all* possible choices of the base.

**Comment by SansPapyrus683** [comment]
yeah the difference is just how many test cases each solution fails on
with enough rng even the double hash could fail

**Comment by SansPapyrus683** [comment]
https://oj.uz/submission/659670
lol, with double hashing it's too slow

**Review by dongliu0426 — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
should i still add the double hashing solution w/ a disclaimer?

**Comment by bqi343** [comment]
Try to make it faster?

**Comment by dongliu0426** [comment]
https://oj.uz/submission/659934 works with unordered_map

**Comment by SansPapyrus683** [comment]
bre I thought unordered was slower

**Comment by SansPapyrus683** [comment]
https://cp-algorithms.com/string/string-hashing.html#improve-no-collision-probability
so now w/ double hashing it's 10^-6?

**Comment by bqi343** [comment]
> Remember, the probability that collision happens is only  $\approx \frac{1}{m}$ .

hm actually the best upper bound I know how to show is $\frac{n}{m}$. so maybe there are test cases where the expected number of collisions is up to $n^2\cdot (n/m)^2$ (or maybe not, idk).

If you want to provably reduce the expected number of collisions to $n^2/m^2$ then you should use random numbers instead of powers of the same base for `elem_pow`.

**Comment by SansPapyrus683** [comment]
so right now there's still a high chance of collisions?

**Comment by bqi343** [comment]
I do not know how to show that the chance of a collision is low

**Comment by SansPapyrus683** [comment]
i already used double hashing, so what else do you suggest?

**Comment by bqi343** [comment]
> If you want to provably reduce the expected number of collisions to  then you should use random numbers instead of powers of the same base for elem_pow.

this

**Comment by SansPapyrus683** [comment]
alright, done

File: solutions/coci-16-zamjene.mdx [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+$$
+
+<Info>
+
+It's also possible to assign each distinct number in the array a distinct power of a prime $P$.
+
+</Info>
+
+Each component should not only have a hash of its number, but also a hash of the numbers it's supposed to have.
+For example, if our component was just the number $3$ but needs a $4$ to be sorted, its hash and required hash would be $P^3$ and $P^4$ respectively.
```
Comment (bqi343) [inline solutions/coci-16-zamjene.mdx:50]: ^ this should be updated?

File: solutions/coci-16-zamjene.mdx [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+			// for everything to be sortable, there can't be any bad components
+			return bad_num == 0;
+		}
+
+		ll needed_pair_num() {
+			return cloud_pairs;
+		}
+};
+
+// https://oj.uz/problem/view/COCI16_zamjene (input omitted due to length)
```
Comment (bqi343) [inline solutions/coci-16-zamjene.mdx:277]: it looks like you undid some of your last few commits

File: solutions/coci-16-zamjene.mdx (hunk 5 above)
Comment (bqi343) [inline solutions/coci-16-zamjene.mdx:50]: ^ this should be updated

File: solutions/coci-16-zamjene.mdx [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+we can obtain our hash by multiplying each random number by the number of times it occurs in the set.
+
+For example, if our component was the multiset ${3, 3, 4, 4, 5}$, and $R_i$ would be the random number of $i$, our hash would be:
+$$
+2 \cdot R_3 + 2 \cdot R_4 + 1 \cdot R_5
+$$
+
+<Info>
+
+It's also possible to assign each distinct number in the array a distinct power of a prime $P$.
```
Comment (bqi343) [inline solutions/coci-16-zamjene.mdx:45]: either remove this or add a note that doing so does not guarantee the same failure probability

File: solutions/coci-16-zamjene.mdx [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+
+This motivates us keeping a map of the difference between the required hash and actual hash and the number of positions whose components have such a difference.
+It doesn't matter whether we subtract the required hash from the actual hash or vice versa, as long as we stay consistent.
+We also keep track of the number of pairs of bad components that can be matched up.
+
+When adding or removing a bad component, we get its difference $d$ and see how many positions whose components have the difference $-d$ through the map to update the number of pairs.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(Q \cdot \alpha(N))$
```
Comment (bqi343) [inline solutions/coci-16-zamjene.mdx:84]: also add the probability of failure below this (and optionally, a short explanation about it)

**Review by bqi343 — CHANGES_REQUESTED** [review]
just a few minor things

**Comment by SansPapyrus683** [comment]
so, we done?

File: solutions/coci-16-zamjene.mdx [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+we can obtain our hash by multiplying each random number by the number of times it occurs in the set.
+
+For example, if our component was the multiset ${3, 3, 4, 4, 5}$, and $R_i$ would be the random number of $i$, our hash would be:
+$$
+2 \cdot R_3 + 2 \cdot R_4 + 1 \cdot R_5
+$$
+
+<Info>
+
+It's also possible to assign each distinct number in the array a distinct power of a prime $P$.
```
Comment (bqi343) [inline solutions/coci-16-zamjene.mdx:45]: actually I"m confused by this, why would you choose P to be prime (as opposed to some random integer in $[0,MOD)$)

Comment (SansPapyrus683) [reply solutions/coci-16-zamjene.mdx:45]: isn't that what you do w/ strings?

Comment (bqi343) [reply solutions/coci-16-zamjene.mdx:45]: no it isn't

Comment (SansPapyrus683) [reply solutions/coci-16-zamjene.mdx:45]: image
what's this

Comment (bqi343) [reply solutions/coci-16-zamjene.mdx:45]: I think that's just wrong, there is no particular reason to make $p$ prime.

Comment (SansPapyrus683) [reply solutions/coci-16-zamjene.mdx:45]: my whole life is a lie

**Pull request merged**